src/Handler/UnionHandler.php

```diff
@@ -47,46 +48,90 @@ public function serializeUnion(
-        return $this->matchSimpleType($data, $type, $context);
-    }
+    ): mixed {
+        if ($this->isPrimitiveType(gettype($data))) {
+            return $this->matchSimpleType($data, $type, $context);
+        } else {
+            $resolvedType = [
+                'name' => get_class($data),
+                'params' => [],
+            ];
 
-    public function deserializeUnion(DeserializationVisitorInterface $visitor, mixed $data, array $type, DeserializationContext $context)
-    {
-        if ($data instanceof \SimpleXMLElement) {
-            throw new RuntimeException('XML deserialisation into union types is not supported yet.');
+            return $context->getNavigator()->accept($data, $resolvedType);
         }
-
-        return $this->matchSimpleType($data, $type, $context);
     }
 
-    private function matchSimpleType(mixed $data, array $type, Context $context)
+    public function deserializeUnion(DeserializationVisitorInterface $visitor, mixed $data, array $type, DeserializationContext $context): mixed
     {
-        $dataType = $this->determineType($data, $type, $context->getFormat());
-        $alternativeName = null;
-
-        if (isset(static::$aliases[$dataType])) {
-            $alternativeName = static::$aliases[$dataType];
+        if ($data instanceof \SimpleXMLElement) {
+            throw new RuntimeException('XML deserialisation into union types is not supported yet.');
         }
 
         foreach ($type['params'] as $possibleType) {
-            if ($possibleType['name'] === $dataType || $possibleType['name'] === $alternativeName) {
-                return $context->getNavigator()->accept($data, $possibleType);
+            $finalType = null;
+
+            if (!$context->getMetadataStack()->isEmpty()) {
+                $propertyMetadata = $context->getMetadataStack()->top();
+                if (null !== $propertyMetadata->unionDiscriminatorField) {
+                    if (!array_key_exists($propertyMetadata->unionDiscriminatorField, $data)) {
+                        throw new NonVisitableTypeException('Union Discriminator Field \'' . $propertyMetadata->unionDiscriminatorField . '\' not found in data');
+                    }
+
+                    $lkup = $data[$propertyMetadata->unionDiscriminatorField];
+                    if (!empty($propertyMetadata->unionDiscriminatorMap)) {
+                        if (array_key_exists($lkup, $propertyMetadata->unionDiscriminatorMap)) {
+                            $finalType = [
+                                'name' => $propertyMetadata->unionDiscriminatorMap[$lkup],
+                                'params' => [],
+                            ];
+                        } else {
+                            throw new NonVisitableTypeException('Union Discriminator Map does not contain key \'' . $lkup . '\'');
+                        }
+                    } else {
+                        $finalType = [
+                            'name' => $lkup,
```

Comment:
i think that is an anti pattern to leak class name into the JSON representation.... it might be a "quick thing" but it has backfired at me so many times...

are you using this approach? (as example in the "inheritance" discriminator map this feature is intentionally requiring the discriminator map to be not empty)

Reply:
I'm not sure I understand the alternative to putting the className into the type.

This can happen in many ways.  The simplest I can think of is:

```
class SomeClass {
    private MappedDiscriminatedAuthor $data
}
```

will result in such a type:
```
$type = ['name' => 'MappedDiscriminatedAuthor', 'params' => []]
```

I think I must be misunderstanding your question.